flocker/common/_retry.py

```diff
@@ -275,3 +279,178 @@ def should_retry(e):
+
+    :param callable decorator: A unary callable that takes a method and returns
+        a method.
+
+    :return: An object like ``obj`` but with all the methods decorated.
+    """
+    return _DecoratedInstance(obj, decorator)
+
+
+def _poll_until_success_returning_result(
+    should_retry, steps, sleep, function, args, kwargs
+):
+    """
+    Call a function until it does not raise an exception or ``should_retry``
+    says it shouldn't be tried anymore, whichever comes first.
+
+    :param should_retry: A three-argument callable which determines whether
+        further retries are attempted.  If ``None`` or a ``timedelta`` is
+        returned, another retry is attempted (immediately or after sleeping for
+        the indicated interval, respectively).  If an exception is raised,
+        further tries are not attempted and the exception is allowed to
+        propagate.
+    :param steps: XXX
+    :param sleep: A function like ``time.sleep`` to use for the delays.
+    :param function: The function to try calling.
+    :param args: Position arguments to pass to the function.
+    :param kwargs: Keyword arguments to pass to the function.
+
+    :return: The value returned by ``function`` on the first call where it
+        returns a value instead of raising an exception.
+    """
+    saved_result = [None]
+
+    def pollable():
+        Message.new(
+            message_type=_TRY_RETRYING,
+        ).write()
+        try:
+            result = function(*args, **kwargs)
+        except Exception as e:
+            # XXX: Should this return bool?
+            saved_result[0] = exc_info()
+            should_retry(*saved_result[0])
+            Message.new(
+                message_type=_TRY_FAILURE,
+                exception=str(e),
+            ).write()
+            return False
+        else:
+            Message.new(
+                message_type=_TRY_SUCCESS,
+                result=result,
+            ).write()
+            saved_result[0] = result
+            return True
+
+    try:
+        poll_until(
+            pollable, (step.total_seconds() for step in steps), sleep=sleep,
+        )
+    except LoopExceeded:
+        # XXX untested
+        thing = saved_result.pop()
+        try:
+            raise thing[0], thing[1], thing[2]
+        finally:
+            del thing
+    else:
+        return saved_result[0]
+
+
+def with_retry(method, should_retry=None, steps=None, sleep=None):
+    """
+    Return a new version of ``method`` that retries.
+
+    :param callable method: A method to retry.
+    :param callable should_retry: A one-argument callable which accepts a
+        three-tuple of exception state and returns a ``timedelta`` or raises
+        an exception. If the call should be retried, the ``timedelta`` gives a
+        time to delay before the retry. If an exception is raised, no further
+        retries are attempted and the exception is propagated from the method
+        call.
```

Comment:
:question: Maybe it's better to change the interface so that instead of passing `should_retry`, callers would just pass the predicate on `exc_value`.